### PR TITLE
Mscroggs/build distributed

### PR DESCRIPTION
**Pull request by mscroggs**
We've exposed the build_distributed_mesh function to python and added a test that uses it as an example. The new test tests the dof permuter in parallel.

### CODE REVIEW

**Comment by garth-wells** [comment]
More a comment on the current design than this change, but `build_distributed_mesh` is a bit weird with the current `Mesh` constructor design. It would be nicer if it just returned partitioned cells and vertices, which are then passed to the `Mesh` constructor.

The issue preventing this at the moment is the ghost handing, see https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfin/mesh/Partitioning.cpp#L683.

**Comment by mscroggs** [comment]
> More a comment on the current design than this change, but `build_distributed_mesh` is a bit weird with the current `Mesh` constructor design. It would be nicer if it just returned partitioned cells and vertices, which are then passed to the `Mesh` constructor.
> 
> The issue preventing this at the moment is the ghost handing, see https://github.com/FEniCS/dolfinx/blob/master/cpp/dolfin/mesh/Partitioning.cpp#L683.

This might be a useful temporary feature, as we need this to test the dof permuter and other mesh stuff (eg gmsh) in parallel

File: python/test/unit/fem/test_dofmap.py [hunk 1]
```diff
@@ -524,6 +523,69 @@ def test_triangle_dof_ordering(space_type):
...
+    ("P", 1), ("P", 2), ("P", 3), ("P", 4),
+    ("N1curl", 1),
+    ("RT", 1), ("RT", 2), ("RT", 3), ("RT", 4),
+    ("BDM", 1),
+    ("N2curl", 1),
+])
+def test_triangle_dof_ordering_parallel(space_type):
+    """Checks that dofs on shared triangle edges match up"""
+    # Create a triangle mesh
+    from random import shuffle
```
Comment (garth-wells) [inline python/test/unit/fem/test_dofmap.py]: Move to the top of the file?

File: python/test/unit/fem/test_dofmap.py [hunk 2]
```diff
@@ -524,6 +523,69 @@ def test_triangle_dof_ordering(space_type):
...
+    N = 10
+    points = np.array([[x / 2, y / 2] for x in range(N) for y in range(N)])
+    cells = []
+    for x in range(N - 1):
```
Comment (garth-wells) [inline python/test/unit/fem/test_dofmap.py]: Add explanation of what this does.

File: python/test/unit/fem/test_dofmap.py [hunk 3]
```diff
@@ -524,6 +523,69 @@ def test_triangle_dof_ordering(space_type):
...
+        for y in range(N - 1):
+            a = N * y + x
+            for cell in [[a, a + 1, a + N + 1], [a, a + N + 1, a + N]]:
+                # Put points in cell in a random order to test dof permuting
+                shuffle(cell)
```
Comment (garth-wells) [inline python/test/unit/fem/test_dofmap.py]: Do we ultimately want to support arbitrary ordering, or insist on anti-clockwise ordering?

File: python/test/unit/fem/test_dofmap.py (hunk 1 above)
Comment (mscroggs) [reply python/test/unit/fem/test_dofmap.py]: done

File: python/test/unit/fem/test_dofmap.py (hunk 2 above)
Comment (mscroggs) [reply python/test/unit/fem/test_dofmap.py]: done

File: python/test/unit/fem/test_dofmap.py (hunk 3 above)
Comment (mscroggs) [reply python/test/unit/fem/test_dofmap.py]: I've changed this to have the cells in anticlockwise order, but the points in the mesh give a random order

**Pull request merged**